README.md
# EarmarkTagCloud

<!--
DO NOT EDIT THIS FILE
It has been generated from a template by Extractly (https://github.com/RobertDober/extractly.git)
and any changes you make in this file will most likely be lost
-->

  An [Earmark](https://github.com/pragdave/earmark) Plugin to easily create tag clouds inside Markdown Documents.

  In its default configuration it translates a list of lines containing keywords with three metric values to html, here
  is a simple example

  If the plugin lines are

      $$ ruby 10 100 4
      $$ elixir 40 800 12

  Earmark would pass in these lines as the `doc` array in the following doctest

      iex> doc = [
      ...> { "ruby 10 100 4", 1},
      ...> { "elixir 40 800 12", 2},
      ...> ]
      ...> EarmarkTagCloud.as_html(doc)
      {[ "<div class=\"earmark-tag-cloud\">\n",
         "  <span style=\"color: #d4d4d4; font-size: 10pt; font-weight: 100;\">ruby</span>\n",
         "  <span style=\"color: #000000; font-size: 40pt; font-weight: 800;\">elixir</span>\n",
         "</div>\n"
      ], []}


  As we can see from the example above the three numeric values above are specifiying

  * font size in pts

  * font weight

  * and a gray scale value between 0 (white) and 12 (black) that matches to 13 gamma corrected
    shades of gray (you can change the settings to more grades, even 50, if you want.
    c.f. Parameterization)

  We can also set parameters like the font-family, or the div-classes

      iex> doc = [
      ...> { "set font-family Times", 1},
      ...> { "set div-classes my-tags", 2},
      ...> { "ruby 10 100 4", 3},
      ...> { "elixir 40 800 12", 4},
      ...> ]
      ...> EarmarkTagCloud.as_html(doc)
      {[ "<div class=\"my-tags\" style=\"font-family: Times;\">\n",
         "  <span style=\"color: #d4d4d4; font-size: 10pt; font-weight: 100;\">ruby</span>\n",
         "  <span style=\"color: #000000; font-size: 40pt; font-weight: 800;\">elixir</span>\n",
         "</div>\n"
      ], []}




## COPYRIGHT & LICENSE

  Apache 2 License

  Copyright © 2016,7,8,9 RobertDober, robert.dober@gmail.com.

  See file `LICENSE` for details.

## Installation

[Available in Hex](https://hex.pm/docs/publish), the package can be installed as:

  1. Add `earmark_tag_cloud` to your list of dependencies in `mix.exs`:

    ```elixir
    def deps do
      [{:earmark_tag_cloud, "~> 0.1.0"}]
    end
    ```

  2. Ensure `earmark_tag_cloud` is started before your application:

    ```elixir
    def application do
      [applications: [:earmark_tag_cloud]]
    end
    ```


CHANGELOG.md
# v0.1.1

* Removed task readme.ex and replaced by [Extractly](https://hex.pm/packages/extractly)  

* Pushed Elixir dep to 1.8

* Coverage


test/support/render.ex
defmodule Support.Render do
  def render_lines(lines, lnb \\ 1), do: lines |> prepare_lines(lnb) |> EarmarkTagCloud.as_html()
  def prepare_lines(lines, lnb \\ 1), do: Enum.zip(lines, (lnb..lnb+Enum.count(lines)))
end


test/acceptance_test.exs
defmodule EarmarkTagCloudTest do
  use ExUnit.Case
  import Support.Render

  describe "no tags" do
    test "empty" do
      lines = []
      assert render_lines(lines) == {["<div class=\"earmark-tag-cloud\">\n", "</div>\n"],[]}
      
    end
    test "settings do not matter" do
      lines = ["set font-family Whatever U want"]
      assert render_lines(lines) == {["<div class=\"earmark-tag-cloud\" style=\"font-family: Whatever U want;\">\n", "</div>\n"],[]}
    end
    # TODO: v0.2 return errors for unsupported keys
    test "illegal keys are not intercepted (in v0.1 at least)" do
      lines = ["set the-answer 42", "set font-family Helvetica"]
      assert render_lines(lines) == {["<div class=\"earmark-tag-cloud\" style=\"font-family: Helvetica;\">\n", "</div>\n"],[]}
    end
  end

  describe "no errors, default settings" do
    test "one tag" do
      lines = [ "elixir rocks 30 900 6"]
      expected = 
      {[ "<div class=\"earmark-tag-cloud\">\n",
        "  <span style=\"color: #bababa; font-size: 30pt; font-weight: 900;\">elixir rocks</span>\n",
        "</div>\n"
      ], []} 

      assert render_lines(lines) == expected
    end
    test "two tags" do
      lines = ["set font-family Arial", "ruby 10 200 3", "elixir 30 400 5"]
      expected = 
      {[ "<div class=\"earmark-tag-cloud\" style=\"font-family: Arial;\">\n",
        "  <span style=\"color: #e0e0e0; font-size: 10pt; font-weight: 200;\">ruby</span>\n",
        "  <span style=\"color: #c8c8c8; font-size: 30pt; font-weight: 400;\">elixir</span>\n",
        "</div>\n"
      ], []} 

      assert render_lines(lines) == expected
    end
  end

  describe "no errors, custom settings" do
    test "set before" do
      lines = ["set scales 50", "ruby 10 200 50", "elixir 30 400 5"]
      expected = 
      {[ "<div class=\"earmark-tag-cloud\">\n",
        "  <span style=\"color: #000000; font-size: 10pt; font-weight: 200;\">ruby</span>\n",
        "  <span style=\"color: #f3f3f3; font-size: 30pt; font-weight: 400;\">elixir</span>\n",
        "</div>\n"
      ], []} 

      assert render_lines(lines) == expected
    end
    test "set in between" do
      lines = [ "ruby 10 200 50", "set scales 50", "elixir 30 400 5"]
      expected = 
      {[ "<div class=\"earmark-tag-cloud\">\n",
        "  <span style=\"color: #000000; font-size: 10pt; font-weight: 200;\">ruby</span>\n",
        "  <span style=\"color: #f3f3f3; font-size: 30pt; font-weight: 400;\">elixir</span>\n",
        "</div>\n"
      ], []} 

      assert render_lines(lines) == expected
    end
    test "set after" do
      lines = ["ruby 10 200 50", " set div-classes alpha", "elixir 30 400 5", "set scales 50", "set div-classes beta and gamma"]
      expected = 
      {[ "<div class=\"beta and gamma\">\n",
        "  <span style=\"color: #000000; font-size: 10pt; font-weight: 200;\">ruby</span>\n",
        "  <span style=\"color: #f3f3f3; font-size: 30pt; font-weight: 400;\">elixir</span>\n",
        "</div>\n"
      ], []} 

      assert render_lines(lines) == expected
    end
  end

  describe "errors and custom settings" do
    test "one complex example should be enough" do
      lines = [
        "set font-family Helvetica U Name It",
        "set scales 10",
        "set div-id hello-world",
        "elixir is king 12 100 3",
        "\\set scales 20",
        "\\set 14 200 48",
        "set scales 50",
        "ruby wont cut it 20 800 51"
      ]
      expected = 
      {[ "<div id=\"hello-world\" class=\"earmark-tag-cloud\" style=\"font-family: Helvetica U Name It;\">\n",
        "  <span style=\"color: #f8f8f8; font-size: 12pt; font-weight: 100;\">elixir is king</span>\n",
        "  <span style=\"color: #3b3b3b; font-size: 14pt; font-weight: 200;\">set</span>\n",
        "</div>\n"
      ], [
        {:error, 89, "missing one or more of necessary integer values (font-size font-weight gray-scale) at end of tag specifcation"},
        {:error, 92, "value for gray-scale out of legal range. Actual: 51, Allowed: 0..50"}
      ]} 

      assert render_lines(lines, 85) == expected
    end

  end
end


test/gamma_correction_test.exs
defmodule GammaCorrectionTest do
  use ExUnit.Case
  import EarmarkTagCloud.GammaCorrection

  doctest EarmarkTagCloud.GammaCorrection
end


mix.exs
defmodule EarmarkTagCloud.Mixfile do
  use Mix.Project

  def project do
    [app: :earmark_tag_cloud,
     version: "0.1.1",
     elixir: "~> 1.8",
     elixirc_paths: elixirc_paths(Mix.env),
     description: description(),
     escript: escript_config(),
     build_embedded: Mix.env == :prod,
     start_permanent: Mix.env == :prod,
     package: package(),
     test_coverage: [tool: ExCoveralls],
     preferred_cli_env: [
       coveralls: :test,
       "coveralls.detail": :test,
       "coveralls.post": :test,
       "coveralls.html": :test
     ],
     deps: deps()]
  end

  # Configuration for the OTP application
  #
  # Type "mix help compile.app" for more information
  def application do
    [applications: [:logger]]
  end

  defp deps do
    [
      {:earmark, "~> 1.3"},
      {:extractly, "~> 0.1", only: :dev},
      {:excoveralls, "~> 0.10.3", only: :test},
      {:ex_doc, ">= 0.0.0", only: :dev},
    ]
  end

  defp description do
    """
    EarmarkTagCloud is a plugin for Earmark allowing for simple Tag Cloud generation.
    """
  end
  defp elixirc_paths(:dev), do: ["lib", "test/support", "examples"]
  defp elixirc_paths(:test), do: ["lib", "test/support", "examples"]
  defp elixirc_paths(_),     do: ["lib"]

  defp escript_config do
    [ main_module: EarmarkTagCloud.CLI ]
  end

  defp package do
    [
      files:       [ "lib", "mix.exs", "README.md", "LICENSE" ],
      maintainers: [
        "Robert Dober <robert.dober@gmail.com>",
      ],
      licenses:    [ "Apache 2 (see the file LICENSE for details)" ],
      links:       %{
        "GitHub" => "https://github.com/RobertDober/earmark_tag_cloud",
      }
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :earmark_tag_cloud, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:earmark_tag_cloud, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/earmark_tag_cloud/cli.ex
defmodule EarmarkTagCloud.CLI do

  def main(argv) do
    argv
    |> parse_args
    |> process
  end

  @cli_options [:div_classes, :gamma, :scales]

  @args """
  usage:

     earmark_tag_cloud --help
     earmark_tag_cloud --version
     earmark_tag_cloud [ options... <file> ]

  convert file from Markdown to HTML.

     where options can be any of:
       --div-classes --gamma --scale 

  """


  defp parse_args(argv) do
    switches = [
      help: :boolean,
      version: :boolean
      ]
    aliases = [
      h: :help,
      v: :version
    ]

    parse = OptionParser.parse(argv, switches: switches, aliases: aliases)
    case  parse  do
      { [ {switch, true } ],  _, _ } -> switch
      { options, [ filename ],  _ }  -> {open_file(filename), options}
      { options, [ ],           _ }  -> {:stdio, options}
      _                              -> :help
    end
  end


  defp process(:help) do
    IO.puts(:stderr, @args)
    IO.puts(:stderr, option_related_help())
  end

  defp process(:version) do
    with {:ok, version} <- :application.get_key(:earmark_tag_cloud, :vsn), do:
      IO.puts( version )
  end

  defp process({io_device, _options}) do
    content = IO.stream(io_device, :line) |> Enum.to_list
    
    {html, errors} = Earmark.as_html(content, Earmark.Plugin.define(EarmarkTagCloud))

    Enum.each(errors, fn {type, ln, msg} ->
      IO.puts(:stderr, "#{type}: lnb: #{ln} #{msg}")
    end)
    IO.puts(html)
  end


  defp open_file(filename), do: io_device(File.open(filename, [:utf8]), filename)

  defp io_device({:ok, io_device}, _), do: io_device
  defp io_device({:error, reason}, filename) do
    IO.puts(:stderr, "#{filename}: #{:file.format_error(reason)}")
    exit(1)
  end

  defp option_related_help do
    @cli_options
    |> Enum.map(&specific_option_help/1)
    |> Enum.join("\n")
  end

  defp specific_option_help( option ) do
    "      --#{option} defaults to #{inspect(Map.get(%Earmark.Options{}, option))}"
  end

end


lib/earmark_tag_cloud/parser.ex
defmodule EarmarkTagCloud.Parser do

   @set_param ~r[^\s*set\s+(\S+)(?:\s+(.*))?]

   @doc """
    Parses a line, which is either a set instruction or a tag spec:

        iex> parse_line({"set font-family Helvetica sans serif", 1})
        {:set, "font-family", "Helvetica sans serif"}

        iex> parse_line({"set font-family", 4})
        {:error, 4, "missing value for variable \\"font-family\\""}

        iex> parse_line({"\\\\set font-family 1 2 3", 40})
        {:tag, "set font-family", [1, 2, 3], 40}

        iex> parse_line({"\\\\set font-family 1 2 ", 42})
        {:error, 42, "missing one or more of necessary integer values (font-size font-weight gray-scale) at end of tag specifcation"}
  """
  def parse_line({line, lnb}) do
    case Regex.run(@set_param, line) do
      [_, key]   -> {:error, lnb, "missing value for variable #{inspect key}"} 
      [_, key, value] -> {:set, key, translate(key, value)}
      _               -> parse_tag({line, lnb})
    end
  end

  @tag_spec ~r'''
    # Tag text
    (.+)
    # Size, Weight, Color
    \s+ (\d+) \s+ (\d+) \s+ (\d+) \s*$
  '''x
  defp parse_tag({line, lnb}) do
    case Regex.run(@tag_spec, line) do
      [ _, text | numbas  ] -> {:tag, Regex.replace(~r{^\\}, text, ""), ints(numbas), lnb}
      _                     -> {:error, lnb, "missing one or more of necessary integer values (font-size font-weight gray-scale) at end of tag specifcation"}
    end
  end

  defp float(numba) do
    with {parsed, _} <- Float.parse(numba), do: parsed
  end

  defp ints(numbas) do 
    numbas |> Enum.map(&int/1)
  end

  defp int(numba) do
    with {parsed, _} <- Integer.parse(numba), do: parsed
  end 

  defp translate(key, value) do
    case key do 
      "scales" -> int(value)
      "gamma" -> float(value)
      _       -> value
    end
  end
end


lib/earmark_tag_cloud/renderer.ex
defmodule EarmarkTagCloud.Renderer do
  
  import EarmarkTagCloud.GammaCorrection
  
  def render(parsed_lines) do 
    parsed_lines
    |> Enum.reduce({[], default_values()}, &remove_and_calc_settings/2)
    |> format_lines({["</div>\n"], []})
  end


  defp format_lines({[],settings},{output_lines, errors}), do: { [div_line(settings)|output_lines], errors }
  defp format_lines({[{:tag, text, params, lnb} | rest], settings}, {output, errors}) do 
    case make_span(text, params, settings) do
      {:error, _, msg} -> format_lines({rest, settings}, {output, [{:error, lnb, msg}|errors]})
      {:ok, span}      -> format_lines({rest, settings}, {[span|output], errors})
    end
  end
  defp format_lines({[err|rest], settings}, {output, errors}) do
    format_lines({rest, settings}, {output, [err|errors]})
  end

  defp remove_and_calc_settings({:set, key, value}, {output, settings}) do
    {output, Map.put(settings, key, value)}
  end 
  defp remove_and_calc_settings({:tag, _, _, _} = tag , {output, settings}) do
    # {make_span(text, values, settings), settings}
    {[tag | output], settings}
  end
  defp remove_and_calc_settings(error, {output, settings}) do
    {[error | output], settings}
  end

  defp div_line(settings) do
    case settings["font-family"] do
      nil -> ~s(<div #{div_id(settings)}class="#{settings["div-classes"]}">\n)
      font_family -> ~s(<div #{div_id(settings)}class="#{settings["div-classes"]}" style="font-family: #{font_family};">\n)
    end
  end
  defp div_id(%{"div-id" => div_id}), do: ~s(id="#{div_id}" )
  defp div_id(_), do: ""

  defp make_span(text, [size, weight, gray_scale], settings) do
    case make_gray(gray_scale, settings) do
      {:error, message} -> {:error, text, message}
      {:ok, gray}       -> {:ok, ~s[  <span style="color: ##{gray}; font-size: #{size}pt; font-weight: #{weight};">#{text}</span>\n]}
    end
  end

  defp default_values, do: %{
    "scales"      => 12,
    "gamma"       => 2.2,
    "div-classes" => "earmark-tag-cloud",
    "font-family" => nil
  }
end


lib/earmark_tag_cloud/gamma_correction.ex
defmodule EarmarkTagCloud.GammaCorrection do
  
  @doc """
    Make an RGB string for the given gray_scale of a number of scales and a given gamma correction
    according to this formula:

        255 * ( gray_scale / scales ) ^ ( 1 / gamma )

    Gamma defaults to 2.2 and you should only change it if the results are not satisfactory.

    See the following [Wikipedia article](https://en.wikipedia.org/wiki/Gamma_correction) for more information.

    Therefore

        iex> make_gray(99, %{"scales" => 100, "gamma" => 2.2})
        {:ok, "1f1f1f"}

        iex> make_gray(0, %{"scales" => 100, "gamma" => 2.2})
        {:ok, "ffffff"}

        iex> make_gray(10, %{"scales" => 12, "gamma" => 2.2})
        {:ok, "717171"}

        iex> make_gray(13, %{"scales" => 12, "gamma" => 2.2})
        {:error, "value for gray-scale out of legal range. Actual: 13, Allowed: 0..12"}

     Or, if you must

        iex> make_gray(42, %{"scales" => 99, "gamma" => 1.9})
        {:ok, "bfbfbf"}
   
  """
  def make_gray(gray_scale, %{"scales" => scales}) when gray_scale > scales do
    {:error, "value for gray-scale out of legal range. Actual: #{gray_scale}, Allowed: 0..#{scales}"}
  end 
  def make_gray(gray_scale, %{"scales" => scales, "gamma" => gamma}) do
    gamma_correct((scales - gray_scale) / scales, gamma)
  end

  
  defp gamma_correct(gray, gamma) do
    with rgbf = 255 * :math.pow(gray, 1 / gamma) do
      {:ok, rgbf
        |> round()
        |> Integer.to_string(16)
        |> String.pad_leading(2, "00")
        |> String.duplicate(3)
        |> String.downcase()
      }
    end
  end
end


lib/earmark_tag_cloud.ex
defmodule EarmarkTagCloud do

  @moduledoc """
    An [Earmark](https://github.com/pragdave/earmark) Plugin to easily create tag clouds inside Markdown Documents.

    In its default configuration it translates a list of lines containing keywords with three metric values to html, here
    is a simple example

    If the plugin lines are

        $$ ruby 10 100 4
        $$ elixir 40 800 12

    Earmark would pass in these lines as the `doc` array in the following doctest

        iex> doc = [
        ...> { "ruby 10 100 4", 1},
        ...> { "elixir 40 800 12", 2},
        ...> ]
        ...> EarmarkTagCloud.as_html(doc)
        {[ "<div class=\\"earmark-tag-cloud\\">\\n",
           "  <span style=\\"color: #d4d4d4; font-size: 10pt; font-weight: 100;\\">ruby</span>\\n",
           "  <span style=\\"color: #000000; font-size: 40pt; font-weight: 800;\\">elixir</span>\\n",
           "</div>\\n"
        ], []}


    As we can see from the example above the three numeric values above are specifiying

    * font size in pts

    * font weight

    * and a gray scale value between 0 (white) and 12 (black) that matches to 13 gamma corrected
      shades of gray (you can change the settings to more grades, even 50, if you want.
      c.f. Parameterization)

    We can also set parameters like the font-family, or the div-classes

        iex> doc = [
        ...> { "set font-family Times", 1},
        ...> { "set div-classes my-tags", 2},
        ...> { "ruby 10 100 4", 3},
        ...> { "elixir 40 800 12", 4},
        ...> ]
        ...> EarmarkTagCloud.as_html(doc)
        {[ "<div class=\\"my-tags\\" style=\\"font-family: Times;\\">\\n",
           "  <span style=\\"color: #d4d4d4; font-size: 10pt; font-weight: 100;\\">ruby</span>\\n",
           "  <span style=\\"color: #000000; font-size: 40pt; font-weight: 800;\\">elixir</span>\\n",
           "</div>\\n"
        ], []}


  """
  def as_html(lines) do
    lines
    |> Enum.map(&EarmarkTagCloud.Parser.parse_line/1)
    |> EarmarkTagCloud.Renderer.render()
  end
end
